### PR TITLE
SALTO-4990: Salesforce: Change `RecordType` picklist values to references

### PR DESCRIPTION
**Pull request by Nurdok**
This is the second part in remodeling picklists to ordered maps.
This PR handles picklists in `RecordType` objects and replaces their subset of picklist values to references.

Here's a workspace diff: https://github.com/salto-io/amir-sf1/commit/b9cba6ce02cbeaf4dbb0a8ecd793406899a712d6#diff-07dbed1f2950e16a3d8c327218d0840daf20f7188707e551a97579ad2d57933a
Note that since this uses the same FF as the previous PR, the diff includes the ordered map change in Picklist definitions. The link above is a direct link to the file affected by _this_ PR.

---

_Additional context for reviewer_:
None

---
_Release Notes_: 
- `RecordType` picklist values are now represented as references to the original picklist values (protected by `picklistsAsMaps` feature flag).

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/salesforce-adapter/src/filters/picklist_references.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      .filter(objectType => objectType.elemID.typeName === 'RecordType')
+    const picklistValuesItem = recordTypes.flatMap(rt => rt.value.picklistValues).filter(isDefined)
+    picklistValuesItem.forEach(picklistValues => {
+      const picklistRef: ReferenceExpression | undefined =
+        // Some picklist references are themselves references to value sets, while others are direct picklists references.
+        picklistValues.picklist?.value?.annotations?.valueSetName ?? picklistValues.picklist
+      if (!isReferenceExpression(picklistRef)) {
+        return
+      }
+      picklistValues.values = picklistValues.values.map((value: { fullName: string | undefined }) => {
```
Comment (yelly) [inline packages/salesforce-adapter/src/filters/picklist_references.ts:39]: Nit: I prefer:
```suggestion
      picklistValues.values = picklistValues.values.map(({fullName}: { fullName?: string }) => {
```

Comment (yelly) [inline packages/salesforce-adapter/src/filters/picklist_references.ts:39]: Aren't we losing the `default` value here? Is that OK?

File: packages/salesforce-adapter/src/filters/picklist_references.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        if (value.fullName === undefined) {
+          throw new Error('Failed to find value name')
```
Comment (yelly) [inline packages/salesforce-adapter/src/filters/picklist_references.ts:41]: Throwing seems a bit violent here, maybe just log an error and move on.

Comment (yelly) [reply packages/salesforce-adapter/src/filters/picklist_references.ts:41]: Then you can also cover this case without it being too much of a headache.

**Comment by yelly** [comment]
Please add a link to a commit which compares a workspace before and after the feature flag is enabled.
Example: https://github.com/salto-io/salto/pull/6285

File: packages/salesforce-adapter/src/filters/picklist_references.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+ *
+ * @param picklistValues    The picklistValues of a RecordType instance to modify
+ */
+const resolvePicklistValueReferences = (picklistValues: PicklistValuesItem): void => {
+  if (picklistValues.values.some(value => value.fullName !== undefined || value.value === undefined)) {
+    return
+  }
+
+  picklistValues.values = picklistValues.values.map(({ value, ...rest }) => ({
+    fullName: (value as { fullName: string }).fullName,
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/picklist_references.ts:102]: Lets avoid the casting here 🙏

File: packages/salesforce-adapter/src/filters/picklist_references.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  return {
+    name: 'picklistReferences',
+    onFetch: async elements =>
+      elements
+        .filter(isInstanceOfTypeSync(RECORD_TYPE_METADATA_TYPE))
+        .flatMap(rt => rt.value.picklistValues.map((picklistValueItem: PicklistValuesItem) => [rt, picklistValueItem]))
+        .filter(isDefined)
+        .forEach(([rt, pvi]) => addPicklistValueReferences(rt, pvi)),
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/filters/picklist_references.ts:124]: Here too, let's avoid short names

**Review by tamtamirr — COMMENTED** [review]
LGTM! Added some small comments

File: packages/salesforce-adapter/src/filters/picklist_references.ts (hunk 3 above)
Comment (Nurdok) [reply packages/salesforce-adapter/src/filters/picklist_references.ts]: The file looks outdated, please make sure you're looking at the latest version

File: packages/salesforce-adapter/jest.config.js [hunk 5]
```diff
@@ -16,7 +16,7 @@ module.exports = deepMerge(require('../../jest.base.config.js'), {
     : undefined,
   coverageThreshold: {
     global: {
-      branches: 87.75,
+      branches: 87.53,
```
Comment (tamtamirr) [inline packages/salesforce-adapter/jest.config.js:19]: Why is the decrease in coverage?

**Review by tamtamirr — APPROVED** [review]
LGTM! Approved with one small comment

File: packages/salesforce-adapter/jest.config.js (hunk 5 above)
Comment (Nurdok) [reply packages/salesforce-adapter/jest.config.js:19]: For some reason it claims ~50% branch coverage, though AFAIK I covered all the conditionals. Doesn't seem to worth the hassle to increase this.

**Pull request merged**